compiler/damlc/daml-doc/src/DA/Daml/Doc/Driver.hs

```diff
@@ -18,67 +21,99 @@ import Development.IDE.Types.Location
-import           System.Directory
-import           System.FilePath
 
+data DamldocOptions = DamldocOptions
+    { do_inputFormat :: InputFormat
+    , do_ideOptions :: IdeOptions
+    , do_outputPath :: FilePath
+    , do_outputFormat :: OutputFormat
+    , do_docTemplate :: Maybe FilePath
+    , do_docOptions :: [DocOption]
+    , do_inputFiles :: [NormalizedFilePath]
+    , do_docTitle :: Maybe T.Text
+    , do_combine :: Bool
+    }
 
 data InputFormat = InputJson | InputDaml
     deriving (Eq, Show, Read)
 
-damlDocDriver :: InputFormat
-              -> IdeOptions
-              -> FilePath
-              -> DocFormat
-              -> Maybe FilePath
-              -> [DocOption]
-              -> [NormalizedFilePath]
-              -> IO ()
-damlDocDriver cInputFormat ideOpts output cFormat prefixFile options files = do
+data OutputFormat = OutputJson | OutputHoogle | OutputDocs RenderFormat
+    deriving (Eq, Show, Read)
+
+-- | Run damldocs!
+damlDocDriver :: DamldocOptions -> IO ()
+damlDocDriver options = do
+    docData <- inputDocData options
+    renderDocData options (transformDocData options docData)
+
+-- | Load doc data, either via the DAML typechecker or via JSON files.
+inputDocData :: DamldocOptions -> IO [ModuleDoc]
+inputDocData DamldocOptions{..} = do
 
     let printAndExit errMsg = do
-            putStrLn $ "Error processing input from " <> unwords (map fromNormalizedFilePath files) <> "\n"
-                     <> errMsg
-            fail "Aborted."
+            hPutStr stderr . unlines $
+                [ unwords
+                    $ "Error processing input from"
+                    : map fromNormalizedFilePath do_inputFiles
+                , errMsg
+                ]
+            exitFailure
 
-    let onErrorExit act =
-            act >>= either (printAndExit . renderDiags) pure
         renderDiags = T.unpack . showDiagnosticsColored
+        onErrorExit act = act >>= either (printAndExit . renderDiags) pure
+
+    case do_inputFormat of
+        InputJson -> do
+            input <- mapM (BS.readFile . fromNormalizedFilePath) do_inputFiles
+            let mbData = map (AE.eitherDecode . LBS.fromStrict) input
+            applyAnnotations <$> concatMapM (either printAndExit pure) mbData
 
-    docData <- case cInputFormat of
-            InputJson -> do
-                input <- mapM (BS.readFile . fromNormalizedFilePath) files
-                let mbData = map AE.eitherDecode input :: [Either String [ModuleDoc]]
-                applyAnnotations <$> concatMapM (either printAndExit pure) mbData
+        InputDaml -> onErrorExit . runExceptT $
+            fmap (applyTransform do_docOptions)
+                (mkDocs do_ideOptions do_inputFiles)
 
-            InputDaml ->
-                onErrorExit $ runExceptT
-                            $ fmap (applyTransform options)
-                            $ mkDocs ideOpts files
+-- | Transform doc data, applying annotations and
```

Comment:
Looks like that comment stops in the middle of a sentence. Given that this only seems to be used once and I don’t find either the name or the comment to add much helpful information, I would just inline the implementation.

Reply:
Fair